File: firmware/buttons.go
//go:build tinygo && !modemonly

package main

import (
	"machine"
	"runtime/volatile"
	"time"
)

var (
	btnPendNext volatile.Register8
	btnPendOK   volatile.Register8
)

type Buttons struct {
	next, ok           machine.Pin
	lastNext, lastOK   bool // true = high / отпущена (pull-up)
	nextAt, okAt       time.Time
	debounce           time.Duration
}

func NewButtons() *Buttons {
	b := &Buttons{
		next:     pinBtnNext,
		ok:       pinBtnOK,
		debounce: 40 * time.Millisecond,
		lastNext: true,
		lastOK:   true,
	}
	b.next.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
	b.ok.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
	b.lastNext = b.next.Get()
	b.lastOK = b.ok.Get()

	// IRQ — бонус; основной путь — опрос уровня (WiFi может сбрасывать INTENABLE).
	_ = b.next.SetInterrupt(machine.PinFalling, func(machine.Pin) {
		btnPendNext.Set(1)
	})
	_ = b.ok.SetInterrupt(machine.PinFalling, func(machine.Pin) {
		btnPendOK.Set(1)
	})
	return b
}

func (b *Buttons) Pending() bool {
	if btnPendNext.Get() != 0 || btnPendOK.Get() != 0 {
		return true
	}
	// зажата прямо сейчас
	return !b.next.Get() || !b.ok.Get()
}

// Poll: active-low на GND. IRQ-флаг или фронт high→low по Get().
func (b *Buttons) Poll() (pressedNext, pressedOK bool) {
	now := time.Now()

	if btnPendNext.Get() != 0 {
		btnPendNext.Set(0)
		if now.Sub(b.nextAt) > b.debounce {
			pressedNext = true
			b.nextAt = now
			b.lastNext = false
		}
	}
	if btnPendOK.Get() != 0 {
		btnPendOK.Set(0)
		if now.Sub(b.okAt) > b.debounce {
			pressedOK = true
			b.okAt = now
			b.lastOK = false
		}
	}

	n := b.next.Get()
	o := b.ok.Get()
	if !pressedNext && !n && b.lastNext && now.Sub(b.nextAt) > b.debounce {
		pressedNext = true
		b.nextAt = now
	}
	if !pressedOK && !o && b.lastOK && now.Sub(b.okAt) > b.debounce {
		pressedOK = true
		b.okAt = now
	}
	b.lastNext = n
	b.lastOK = o
	return
}

File: firmware/src/buttons.cpp
#include "buttons.h"
#include "pins.h"

volatile uint8_t g_btn_pend_next = 0;
volatile uint8_t g_btn_pend_ok = 0;

static void IRAM_ATTR onNextISR() { g_btn_pend_next = 1; }
static void IRAM_ATTR onOkISR() { g_btn_pend_ok = 1; }

void Buttons::begin() {
  pinMode(PIN_BTN_NEXT, INPUT_PULLUP);
  pinMode(PIN_BTN_OK, INPUT_PULLUP);
  last_next_ = digitalRead(PIN_BTN_NEXT);
  last_ok_ = digitalRead(PIN_BTN_OK);
  attachInterrupt(digitalPinToInterrupt(PIN_BTN_NEXT), onNextISR, FALLING);
  attachInterrupt(digitalPinToInterrupt(PIN_BTN_OK), onOkISR, FALLING);
}

bool Buttons::pending() const {
  if (g_btn_pend_next || g_btn_pend_ok) return true;
  return digitalRead(PIN_BTN_NEXT) == LOW || digitalRead(PIN_BTN_OK) == LOW;
}

bool Buttons::nextDown() const { return digitalRead(PIN_BTN_NEXT) == LOW; }
bool Buttons::okDown() const { return digitalRead(PIN_BTN_OK) == LOW; }

void Buttons::poll(bool *pressed_next, bool *pressed_ok) {
  *pressed_next = false;
  *pressed_ok = false;
  uint32_t now = millis();

  if (g_btn_pend_next) {
    g_btn_pend_next = 0;
    if (now - next_at_ > debounce_ms_) {
      *pressed_next = true;
      next_at_ = now;
      last_next_ = false;
    }
  }
  if (g_btn_pend_ok) {
    g_btn_pend_ok = 0;
    if (now - ok_at_ > debounce_ms_) {
      *pressed_ok = true;
      ok_at_ = now;
      last_ok_ = false;
    }
  }

  bool n = digitalRead(PIN_BTN_NEXT);
  bool o = digitalRead(PIN_BTN_OK);
  if (!*pressed_next && !n && last_next_ && now - next_at_ > debounce_ms_) {
    *pressed_next = true;
    next_at_ = now;
  }
  if (!*pressed_ok && !o && last_ok_ && now - ok_at_ > debounce_ms_) {
    *pressed_ok = true;
    ok_at_ = now;
  }
  last_next_ = n;
  last_ok_ = o;
}
